pct.sh

```diff
@@ -18,23 +18,99 @@ if [[ -n ${EXTRA_MAVEN_PROPERTIES-} ]]; then
 fi
 
 #
-# Plugin Compatibility Tester (PCT) requires custom --add-opens directives when running on Java 17.
-# As a temporary workaround, we pass in these directives when invoking PCT. When
-# jenkinsci/plugin-compat-tester#352 is merged and released, and when this repository has upgraded
-# to that release, this workaround can be deleted.
+# Grab the Jenkins version from the WAR file so that we can pass it in via jenkins.version. This is
+# needed because HPI Plugin requires the version of the WAR passed in via overrideWar to be
+# identical to jenkins.version. If we do not explicitly pass in jenkins.version, then the
+# jenkins.version defined in the plugin's pom.xml file will be used, which may not match the version
+# of the WAR under test.
 #
+mkdir pct-work
+pushd pct-work
+jar xf ../megawar.war META-INF/MANIFEST.MF
+JENKINS_VERSION=$(grep Jenkins-Version META-INF/MANIFEST.MF | awk '{print $2}')
+popd
+rm -rf pct-work
+MAVEN_PROPERTIES+=":jenkins.version=${JENKINS_VERSION}:overrideWar=$(pwd)/megawar.war:useUpperBounds=true"
```

Comment:
The main substance of this change: adding `overrideWar` from [JENKINS-45047](https://issues.jenkins.io/browse/JENKINS-45047), along with its prerequisite `jenkins.version`.

The use of its companion option `useUpperBounds` in this context does not usually affect plugin versions, as would often be the case in the original use case described in JENKINS-45047. This is because the entire tree of plugin versions is already specified in the BOM and therefore the megawar, so there should not be any upper bounds errors here in the common case. However, this option is effective at dealing with some mundane technical issues, like aligning the version of SpotBugs annotations used between core and some library used by a plugin. It might also be effective at helping deal with libraries that have been detached into plugins, although in at least the case of Jakarta Mail it was not effective for this purpose due to the Jakarta Mail maintainers using the same group ID and artifact ID for multiple releases of their plugin, one with `javax` packages and one with `jakarta` packages. If you observe the actual upper bounds errors fixed in a test run, they all represent TODOs that would be "nice to fix" in most cases (e.g. SpotBugs, Joda Time, etc) but not serious enough to warrant the effort in many cases, and it is better to deal with them automatically rather than to allow these mundane issues to block a test run.